internal/ctrlflow/ctrlflow.go

```diff
@@ -0,0 +1,184 @@
+
+	defaultSplit  = 0
+	defaultJunk   = 0
+	defaultPasses = 1
+)
+
+type directiveParamMap map[string]string
+
+func (m directiveParamMap) GetInt(name string, def int) int {
+	rawVal, ok := m[name]
+	if !ok {
+		return def
+	}
+
+	val, err := strconv.Atoi(rawVal)
+	if err != nil {
+		panic(fmt.Errorf("invalid flag %s format: %v", name, err))
+	}
+	return val
+}
+
+// parseDirective parses a directive string and returns a map of directive parameters.
+// Each parameter should be in the form "key=value" or "key"
+func parseDirective(directive string) (directiveParamMap, bool) {
+	if !strings.HasPrefix(directive, directiveName) {
+		return nil, false
+	}
+
+	fields := strings.Fields(directive)
+	if len(fields) <= 1 {
+		return nil, true
+	}
+	m := make(map[string]string)
+	for _, v := range fields[1:] {
+		params := strings.SplitN(v, "=", 2)
+		if len(params) == 2 {
+			m[params[0]] = params[1]
+		} else {
+			m[params[0]] = ""
+		}
+	}
+	return m, true
+}
+
+// Obfuscate obfuscates control flow of all functions with directive using control flattening.
+// All obfuscated functions are removed from the original file and moved to the new one.
+// Obfuscation can be customized by passing parameters from the directive, example:
+//
+// //garble:controlflow passes=1 junk=0 split=0
+// func someMethod() {}
+//
+// passes - controls number of passes of control flow flattening. Have exponential complexity and more than 3 passes are not recommended in most cases.
+// junk - controls how many junk jumps are added. It does not affect final binary by itself, but together with flattening linearly increases complexity.
+// split - controls number of times largest block must be splitted. Together with flattening improves obfuscation of long blocks without branches.
+func Obfuscate(fset *token.FileSet, ssaPkg *ssa.Package, files []*ast.File, obfRand *mathrand.Rand) (newFileName string, newFile *ast.File, affectedFiles []*ast.File, err error) {
+	var ssaFuncs []*ssa.Function
+	var ssaParams []directiveParamMap
+
+	for _, file := range files {
+		affected := false
+		for _, decl := range file.Decls {
+			funcDecl, ok := decl.(*ast.FuncDecl)
+			if !ok || funcDecl.Doc == nil {
+				continue
+			}
+
+			for _, comment := range funcDecl.Doc.List {
+				params, hasDirective := parseDirective(comment.Text)
+				if !hasDirective {
+					continue
+				}
+
+				path, _ := astutil.PathEnclosingInterval(file, funcDecl.Pos(), funcDecl.Pos())
+				ssaFunc := ssa.EnclosingFunction(ssaPkg, path)
+				if ssaFunc == nil {
+					panic("function exists in ast but not found in ssa")
+				}
+
+				ssaFuncs = append(ssaFuncs, ssaFunc)
+				ssaParams = append(ssaParams, params)
+
+				log.Printf("detected function for controlflow %s (params: %v)", funcDecl.Name.Name, params)
+
+				// Remove inplace function from original file
+				funcDecl.Name = ast.NewIdent("_")
+				funcDecl.Body = ah.BlockStmt()
+				funcDecl.Recv = nil
+				funcDecl.Type = &ast.FuncType{Params: &ast.FieldList{}}
```

Comment:
perhaps add a TODO to remove the func declaration entirely, that sort of thing is easy to do with `astutil.Cursor`, or could likely be done by replacing the `Decls` slice yourself.